File: sdks/swift/Sources/Chio/ReceiptPoster.swift
import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public enum MobileReceiptErrorCode: String, Codable, Equatable {
    case receiptPostRejected = "urn:chio:error:mobile:receipt-post-rejected"
    case receiptSchemaInvalid = "urn:chio:error:mobile:receipt-schema-invalid"
    case oracleUnreachable = "urn:chio:error:mobile:oracle-unreachable"
}

public struct ReceiptPostResult: Equatable {
    public let receiptHash: String
    public let accepted: Bool
    public let statusCode: Int?
    public let errorCode: MobileReceiptErrorCode?
}

public final class ReceiptPoster {
    private let endpoint: URL
    private let session: URLSession
    private let maxAttempts: Int
    private let baseDelayNanoseconds: UInt64

    public init(
        endpoint: URL,
        session: URLSession = .shared,
        maxAttempts: Int = 4,
        baseDelayNanoseconds: UInt64 = 50_000_000
    ) {
        self.endpoint = endpoint
        self.session = session
        self.maxAttempts = max(1, maxAttempts)
        self.baseDelayNanoseconds = baseDelayNanoseconds
    }

    public func flush(queue: OfflineReceiptQueue) async -> [ReceiptPostResult] {
        let queued: [QueuedMobileReceipt]
        do {
            queued = try queue.oldestFirst()
        } catch {
            return [
                ReceiptPostResult(
                    receiptHash: "offline-queue",
                    accepted: false,
                    statusCode: nil,
                    errorCode: .oracleUnreachable
                ),
            ]
        }

        var results: [ReceiptPostResult] = []
        for receipt in queued {
            let result = await post(receipt)
            results.append(result)
            if result.accepted {
                try? queue.remove(tenantId: receipt.tenantId, receiptHash: receipt.receiptHash)
            } else {
                try? queue.markAttempt(receipt)
            }
        }
        return results
    }

    public func post(_ receipt: QueuedMobileReceipt) async -> ReceiptPostResult {
        for attempt in 1...maxAttempts {
            var request = URLRequest(url: endpoint)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = receipt.receiptJson.data(using: .utf8)

            do {
                let (_, response) = try await session.data(for: request)
                let statusCode = (response as? HTTPURLResponse)?.statusCode
                if let statusCode, (200..<300).contains(statusCode) {
                    return ReceiptPostResult(
                        receiptHash: receipt.receiptHash,
                        accepted: true,
                        statusCode: statusCode,
                        errorCode: nil
                    )
                }
                if statusCode == 400 || statusCode == 422 {
                    return rejected(receipt, statusCode: statusCode, code: .receiptSchemaInvalid)
                }
                if let statusCode, [401, 403, 409].contains(statusCode) {
                    return rejected(receipt, statusCode: statusCode, code: .receiptPostRejected)
                }
            } catch {
                if attempt == maxAttempts {
                    return rejected(receipt, statusCode: nil, code: .oracleUnreachable)
                }
            }

            if attempt < maxAttempts {
                let multiplier = UInt64(1 << UInt64(attempt - 1))
                try? await Task.sleep(nanoseconds: baseDelayNanoseconds * multiplier)
            }
        }

        return rejected(receipt, statusCode: nil, code: .oracleUnreachable)
    }

    private func rejected(
        _ receipt: QueuedMobileReceipt,
        statusCode: Int?,
        code: MobileReceiptErrorCode
    ) -> ReceiptPostResult {
        ReceiptPostResult(
            receiptHash: receipt.receiptHash,
            accepted: false,
            statusCode: statusCode,
            errorCode: code
        )
    }
}

File: sdks/jvm/chio-kernel-mobile/src/main/kotlin/dev/chio/kernel/ReceiptPoster.kt
package dev.chio.kernel

import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import kotlin.math.max

enum class MobileReceiptErrorCode(val urn: String) {
    RECEIPT_POST_REJECTED("urn:chio:error:mobile:receipt-post-rejected"),
    RECEIPT_SCHEMA_INVALID("urn:chio:error:mobile:receipt-schema-invalid"),
    ORACLE_UNREACHABLE("urn:chio:error:mobile:oracle-unreachable"),
}

data class ReceiptPostResult(
    val receiptHash: String,
    val accepted: Boolean,
    val statusCode: Int?,
    val errorCode: MobileReceiptErrorCode?,
)

class ReceiptPoster(
    private val endpoint: URL,
    private val maxAttempts: Int = 4,
    private val baseDelayMillis: Long = 50,
) {
    fun flush(queue: OfflineReceiptQueue): List<ReceiptPostResult> =
        queue.oldestFirst().map { receipt ->
            val result = post(receipt)
            if (result.accepted) {
                queue.remove(receipt.tenantId, receipt.receiptHash)
            } else {
                queue.markAttempt(receipt)
            }
            result
        }

    fun post(receipt: QueuedMobileReceipt): ReceiptPostResult {
        val attempts = max(1, maxAttempts)
        for (attempt in 1..attempts) {
            try {
                val status = postOnce(receipt)
                if (status in 200..299) {
                    return accepted(receipt, status)
                }
                if (status == 400 || status == 422) {
                    return rejected(receipt, status, MobileReceiptErrorCode.RECEIPT_SCHEMA_INVALID)
                }
                if (status == 401 || status == 403 || status == 409) {
                    return rejected(receipt, status, MobileReceiptErrorCode.RECEIPT_POST_REJECTED)
                }
            } catch (_: IOException) {
                if (attempt == attempts) {
                    return rejected(receipt, null, MobileReceiptErrorCode.ORACLE_UNREACHABLE)
                }
            }

            if (attempt < attempts) {
                Thread.sleep(baseDelayMillis * (1L shl (attempt - 1)))
            }
        }
        return rejected(receipt, null, MobileReceiptErrorCode.ORACLE_UNREACHABLE)
    }

    private fun postOnce(receipt: QueuedMobileReceipt): Int {
        val connection = endpoint.openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.doOutput = true
        connection.setRequestProperty("Content-Type", "application/json")
        connection.outputStream.use { output ->
            output.write(receipt.receiptJson.toByteArray(Charsets.UTF_8))
        }
        return connection.responseCode
    }

    private fun accepted(receipt: QueuedMobileReceipt, statusCode: Int): ReceiptPostResult =
        ReceiptPostResult(
            receiptHash = receipt.receiptHash,
            accepted = true,
            statusCode = statusCode,
            errorCode = null,
        )

    private fun rejected(
        receipt: QueuedMobileReceipt,
        statusCode: Int?,
        code: MobileReceiptErrorCode,
    ): ReceiptPostResult =
        ReceiptPostResult(
            receiptHash = receipt.receiptHash,
            accepted = false,
            statusCode = statusCode,
            errorCode = code,
        )
}
